Enforce txn offset commit authorization

robobario opened the pull request.

robobario committed:
Enforce TxnOffsetCommit topic authorization

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

robobario committed:
Formatting

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

tombentley commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/filter/authorization/TxnOffsetCommitAuthzIT.java:
```diff
@@ -0,0 +1,422 @@
+            adminClientUnproxied = driveTransactionToPointOfTxnOffsetCommit(kafkaClusterWithAuthz, true);
+            adminClientProxied = driveTransactionToPointOfTxnOffsetCommit(kafkaClusterNoAuthz, false);
+        }
+        catch (Exception e) {
```
Should be find to declare this in a `throws` clause of the method declaration.

robobario replied:
removed in the refactor

tombentley commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/filter/authorization/TxnOffsetCommitAuthzIT.java:
```diff
@@ -0,0 +1,422 @@
+
+    }
+
+    private static MemberIdAndGeneration joinGroup(KafkaClient client, String groupId) {
```
Was there a reason why you could have reused the join group stuff I wrote for `OFFSET_COMMIT`?

robobario replied:
haven't looked at that test, will try and get some re-use going

robobario replied:
Have introduced a shared `KafkaDriver` used for the gestures like `findCoordinator` etc.

Still a bit of a mixed bag of how we create the RPCs. You've got a nice multi-version approach and mine are hardcoded to a particular version. I think we should merge and can keep refactoring separately, better to get it merged.

tombentley commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/filter/authorization/TxnOffsetCommitAuthzIT.java:
```diff
@@ -0,0 +1,422 @@
+        });
+    }
+
+    private static void saslPlainAuthenticate(KafkaClient client, String username, String password) {
```
There's code in the base class for authentication. Is there some reason why we can't reuse that?

robobario replied:
have changed to use the same style as the OffsetCommit test, using the authenticated clients.

robobario committed:
Introduce shared KafkaDriver for low-level gestures

Some authorization tests require us to drive the kafka broker into
particular states through complex sequences of RPC calls, before we
make the RPC call we are interested in checking auth for.

For transactional and consumer group relevant RPCs we need to establish
coordinators, initialize producer ids, and other interactions that may
involve polling the broker until resources become available. These
gestures are re-usable across tests.

The KafkaDriver provides us something higher level than the single-RPC
KafkaClient, but retains the control that we don't have with the high
level upstream clients.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>
